components/proxy/src/main/kotlin/com/hotels/styx/routing/db/NotificationQueue.kt

```diff
@@ -0,0 +1,122 @@
+/*
+  Copyright (C) 2013-2019 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.routing.db
+
+import com.hotels.styx.api.configuration.ObjectStore
+import org.pcollections.HashTreePMap
+import org.pcollections.PMap
+import reactor.core.publisher.FluxSink
+import java.util.Optional
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.concurrent.withLock
+
+
+internal class NotificationQueue<T>(val watchers: CopyOnWriteArrayList<ChangeWatcher<T>>, val executor: ExecutorService) {
+    @Volatile
+    private var pendingSnapshot = IndexedSnapshot<T>(0, HashTreePMap.empty())
+    @Volatile
+    private var issuedSnapshot = IndexedSnapshot<T>(0, HashTreePMap.empty())
+    private val pendingChangeNotification = AtomicBoolean(false)
+    private val lock = ReentrantLock()
+
+    private val listeners = ConcurrentHashMap<String, DispatchListener<T>>()
```

Comment:
Are the `listeners` only for use in testing, or is there another purpose?

Reply:
Yes. They are only for testing purposes.